lib/core/models/repositories/roleRepository.js

```diff
@@ -181,25 +195,80 @@ class RoleRepository extends Repository {
    * @returns Promise
    */
   validateAndSaveRole (role, options) {
-    return role.validateDefinition()
+    return role
+      .validateDefinition()
       .then(() => {
         if (role._id === 'anonymous' && !role.canLogIn()) {
           errorsManager.throw('security', 'role', 'login_required');
         }
 
-        this.kuzzle.emit(
-          'core:roleRepository:save',
-          { _id: role._id, controllers: role.controllers });
-
-        return this.persistToDatabase(role, options);
+        this.kuzzle.emit('core:roleRepository:save', {
+          _id: role._id,
+          controllers: role.controllers
+        });
       })
+      .then(() => this.checkRoleControllersAndActions(role))
+      .then(() => this.persistToDatabase(role, options))
       .then(() => this.loadOneFromDatabase(role._id))
       .then(updatedRole => {
         this.roles.set(role._id, updatedRole);
         return updatedRole;
       });
   }
 
+  /**
+   * Given a Role object, checks if its controllers and actions exist.
+   *
+   * @param {Role} role
+   * @returns Promise
+   */
+  checkRoleControllersAndActions(role) {
+    Object.keys(role.controllers).forEach(roleController => {
+      if (roleController === '*') {
+        this.kuzzle.funnel.controllers.forEach(controller => {
+          Object.keys(role.controllers['*'].actions).forEach(action => {
+            if (action !== '*' && !controller._isAction(action)) {
+              return errorsManager.throw(
+                'security',
+                'role',
+                'unknown_controller_action',
+                role._id,
+                action,
+                controller);
+            }
+          });
+        });
+      }
+      else {
+        if (!this.kuzzle.funnel.isNativeController(
```

Comment:
I think this method signature should be modified to take only the controller name as an argument.
Also, if the role define rights for a plugin controller, this method will throw (eg: `s3/upload` ). You should not check anything if the controller isn't a native one